modules/031-linstor/images/linstor-pools-importer/main.go

```diff
@@ -0,0 +1,481 @@
+			if err != nil {
+				klog.Fatalln("failed to generate Kubernetes storage class:", err)
+			}
+			err = kc.Get(ctx, types.NamespacedName{Name: storageClass.GetName()}, &storageClass)
+			if err == nil { // TODO check for 404
+				continue
+			}
+			err = kc.Create(ctx, &storageClass)
+			if err != nil {
+				event := makeKubernetesEvent(&owner, v1.EventTypeWarning, "Failed", "Failed to create Kubernetes storage class"+err.Error())
+				if err := kc.Create(ctx, &event); err != nil {
+					klog.Errorln("Failed to create event", err)
+				}
+				klog.Fatalln("Failed to create Kubernetes storageClass", err)
+			}
+			event := makeKubernetesEvent(&owner, v1.EventTypeNormal, "Created", "Created Kubernetes storage class: "+storageClass.Name)
+			err = kc.Create(ctx, &event)
+			if err != nil {
+				klog.Fatalln("Failed to create event", err)
+			}
+		}
+	}
+}
+
+// Makes loop over storage pool candidates, retruns channel of changed ones
+func runCandidatesLoop(ctx context.Context, f func() ([]StoragePoolCandidate, error), delay time.Duration) <-chan StoragePoolCandidate {
+	ch := make(chan StoragePoolCandidate)
+	var oldCandidates []StoragePoolCandidate
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				close(ch)
+				return
+			default:
+				candidates, err := f()
+				if err != nil {
+					klog.Fatalln("failed to load candidates:", err)
+				}
+			LOOP:
+				for _, candidate := range candidates {
+					name := candidate.GetName()
+					if name == "" {
+						klog.Fatalln("storage pool name can't be empty")
+					}
+					for _, oldCandidate := range oldCandidates {
+						if name == oldCandidate.GetName() {
+							continue LOOP
+						}
+					}
+					ch <- candidate
+				}
+				oldCandidates = candidates
+				time.Sleep(delay)
+			}
+		}
+	}()
+	return ch
+}
+
+// Collects all storage pool candidates from the node
+func getCandidates() ([]StoragePoolCandidate, error) {
+	var candidates []StoragePoolCandidate
+
+	// Getting LVM volume groups
+	var vgs VolumeGroups
+	err := vgs.LoadCandidates()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get LVM volume groups: %s", err)
+	}
+	for _, vg := range vgs {
+		candidates = append(candidates, vg)
+	}
+
+	// Getting LVM thin pools
+	var tps ThinPools
+	err = tps.LoadCandidates()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get LVM thin pools: %s", err)
+	}
+	for _, tp := range tps {
+		candidates = append(candidates, tp)
+	}
+
+	return candidates, nil
+}
+
+type StoragePoolCandidates interface {
+	LoadCandidates() error
+}
+
+// Defines any type of LINSTOR storage pool candidate
+type StoragePoolCandidate interface {
+	GetName() string
+	GetProviderKind() lclient.ProviderKind
+	GetProps() map[string]string
+}
+
+type ThinPools []ThinPool
+type ThinPool struct {
+	Name   string
+	VGName string
+	Tags   []string
+}
+
+func (a *ThinPools) LoadCandidates() error {
+	cmd := exec.Command("lvs", "-oname,vg_name,lv_attr,tags", "--separator=;", "--noheadings", "--config="+lvmConfig)
```

Comment:
`lvs` and `vgs` support JSON output. Maybe it is simpler to use instead custom parser?

Reply:
Thanks for pointing this out.
I just checked that json output format for LVM commands was added in [v2.02.158 - 25th June 2016](https://github.com/lvmteam/lvm2/blob/8dccc2314e2482370bc6e5cf007eb210994abdef/WHATS_NEW#L783) (available since rhel-7.3)

linstor-server itself [still uses](https://github.com/LINBIT/linstor-server/blob/7e646d83dbbadf1ec066e1bc8b29ae018aff1f66/satellite/src/main/java/com/linbit/linstor/layer/storage/lvm/utils/LvmCommands.java#L72-L111) old format. Since we have no much columns, and in terms of compatibility I would like to be stick to the upstream implementation.